pkg/issuer/acme/http/istio.go

```diff
@@ -0,0 +1,201 @@
+	return &spec, nil
+}
+
+func unstructuredSetSpec(unstr *unstructured.Unstructured, spec *map[string]interface{}) error {
+	return unstructured.SetNestedField(unstr.Object, *spec, "spec")
+}
+
+func (s *Solver) ensureIstio(ctx context.Context, ch *cmacme.Challenge, svcName string) (bool *unstructured.Unstructured, err error) {
+	log := logf.FromContext(ctx).WithName("ensureIstio")
+
+	virtualservice, err := s.getVirtualService(ctx, ch)
+	if err != nil {
+		return nil, err
+	}
+
+	if virtualservice == nil {
+		log.Info("creating VirtualService")
+		virtualservice, err = s.createVirtualService(ctx, ch, svcName)
+		if err != nil {
+			return nil, err
+		}
+		log.Info("VirtualService created successfully")
+
+		return virtualservice, nil
+	}
+
+	log.Info("found VirtualService")
+
+	virtualservice, err = s.checkAndUpdateVirtualService(ctx, ch, svcName, virtualservice)
+	if err != nil {
+		return nil, err
+	}
+
+	return virtualservice, nil
+}
+
+func (s *Solver) cleanupVirtualServices(_ context.Context, _ *cmacme.Challenge) error {
+	// Nothing to do, GC will take care of deleting the VirtualServices when the Challenge is deleted
+	return nil
+}
+
+func (s *Solver) getVirtualService(ctx context.Context, ch *cmacme.Challenge) (*unstructured.Unstructured, error) {
+	selector := labels.Set(podLabels(ch)).AsSelector()
+	vsList, err := s.virtualServiceLister.Namespace(ch.Namespace).List(selector)
+	if err != nil {
+		return nil, err
+	}
+	switch len(vsList) {
+	case 0:
+		return nil, nil
+	case 1:
+		return vsList[0], nil
+	default:
+		for _, vs := range vsList[1:] {
+			s.DynamicClient.Resource(util.VirtualServiceGvr()).Namespace(ch.Namespace).Delete(ctx, vs.GetName(), metav1.DeleteOptions{})
+		}
+		return nil, fmt.Errorf("multiple VirtualServices found")
+	}
+}
+
+func (s *Solver) createVirtualService(ctx context.Context, ch *cmacme.Challenge, svcName string) (*unstructured.Unstructured, error) {
+	expectedSpec := createVirtualServiceSpec(ch, svcName)
+
+	vs := newUnstructuredVirtualService(
+		util.VirtualSerivceGvk().Group,
+		util.VirtualSerivceGvk().Version,
+		util.VirtualSerivceGvk().Kind,
+		"cm-acme-http-solver-",
+		ch.Namespace,
+		podLabels(ch),
+		&[]metav1.OwnerReference{*metav1.NewControllerRef(ch, challengeGvk)},
+		expectedSpec,
+	)
+
+	val, err := s.DynamicClient.Resource(util.VirtualServiceGvr()).Namespace(ch.Namespace).Create(ctx, vs, metav1.CreateOptions{})
+	if err != nil {
+		return nil, err
+	}
+	return val, nil
+}
+
+func (s *Solver) checkAndUpdateVirtualService(ctx context.Context, ch *cmacme.Challenge, svcName string, virtualservice *unstructured.Unstructured) (*unstructured.Unstructured, error) {
+	log := logf.FromContext(ctx, "checkAndUpdateVirtualService")
+
+	expectedSpec := createVirtualServiceSpec(ch, svcName)
+
+	spec, err := unstructuredGetSpec(virtualservice)
+	if err != nil {
+		return nil, err
+	}
+	if reflect.DeepEqual(spec, expectedSpec) {
+		return virtualservice, nil
+	}
+
+	log.Info("Updating VirtualService")
```

Comment:
Where possible, log lines should be lowercase with the exception of kinds (VirtualService)